src/main/java/io/strimzi/kafka/bridge/mqtt/mapper/MqttKafkaMapper.java

```diff
@@ -46,113 +33,61 @@ public class MqttKafkaMapper {
      * @see MappingRule
+     * @see MappingResult
      */
-    public String map(String mqttTopic) {
+    public MappingResult map(String mqttTopic) {
+        MappingResult mappingResult = new MappingResult(DEFAULT_KAFKA_TOPIC, null);
+
         for (MappingRule rule : this.rules) {
             Matcher matcher = this.patterns.get(this.rules.indexOf(rule)).matcher(mqttTopic);
-
             if (matcher.matches()) {
-                HashMap<String, String> placeholders = new HashMap<>();
-
                 String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String kafkaKey = rule.getKafkaKey();
 
-                // find MQTT_TOPIC_PLACEHOLDER_REGEX in the kafkaTopicTemplate.
-                Matcher placeholderMatcher = this.placeholderPattern.matcher(rule.getKafkaTopicTemplate());
-                while (placeholderMatcher.find()) {
-                    String placeholderKey = placeholderMatcher.group();
-                    placeholders.put(placeholderKey, null);
+                for (int i = 1; i < matcher.groupCount() + 1; i++) {
+                    mappedKafkaTopic = mappedKafkaTopic.replace("$" + i, matcher.group(i));
+                    kafkaKey = kafkaKey != null ? kafkaKey.replace("$" + i, matcher.group(i)) : null;
                 }
 
-                if (!placeholders.isEmpty()) {
-                    Matcher mqttTopicMatcher = this.placeholderPattern.matcher(rule.getMqttTopicPattern());
-
-                    // find the placeholders in the mqtt topic pattern and assign them a value.
-                    while (mqttTopicMatcher.find()) {
-                        String placeholderKey = mqttTopicMatcher.group();
-                        String placeholderValue = matcher.group(removeBrackets(placeholderKey));
-                        placeholders.put(placeholderKey, placeholderValue);
-                    }
+                // check for pending placeholders replacement in the kafka topic
+                checkPlaceholder(mappedKafkaTopic);
+                // set the mapped kafka topic to the result
+                mappingResult.setKafkaTopic(mappedKafkaTopic);
 
-                    //build the kafka topic using the placeholders.
-                    for (Map.Entry<String, String> entry : placeholders.entrySet()) {
-                        if (entry.getValue() != null) {
-                            mappedKafkaTopic = mappedKafkaTopic.replace(entry.getKey(), entry.getValue());
-                        } else {
-                            throw new IllegalArgumentException("The placeholder " + entry.getKey() + " was not found assigned any value.");
-                        }
-                    }
+                if (kafkaKey != null) {
+                    // check for pending placeholders replacement in the kafka key.
+                    checkPlaceholder(kafkaKey);
+                    // set the mapped kafka key to the result
+                    mappingResult.setKafkaKey(kafkaKey);
```

Comment:
related to the other comment in the `MappingResult` class, can we just create the `MappingResult` passing topic and key in the constructor instead of using setters in two places. Of course, key could be also null but it would be a valid parameter to the constructor.